### PR TITLE
Add pilight-receive datetime filtering.

### PR DESCRIPTION
**Pull request by pilino1234**
This filters the pilight-receive output and removes the JSON objects for the datetime protocol. Invoke pilight-receive with pilight-receive -F to use.

### CODE REVIEW

**Comment by CurlyMoo** [comment]
Why not allow users to add their own protocols in either a whitelist or blacklist.

**Comment by pilino1234** [comment]
Too complicated to program for me :/

**Comment by CurlyMoo** [comment]
Why? pilight has an explode function that can make this real easy. E.g.
```
pilight-receive -F datetime,arctech_switch,conrad_weather
```

```
char **array = NULL;
int n = explode(argv, ",", &array), i = 0;
for(i=0;i<n;i++) {
   printf("%s\n", array[i]);
}
array_free(array[i], n);
```

**Comment by pilino1234** [comment]
I'll have a look if I can get that to work, passing the protocols like that. I though you meant another configuration entry to define protocols for this :smiley:
Where can I find more about the explode function?

**Comment by CurlyMoo** [comment]
What do you need to know more?

**Comment by pilino1234** [comment]
I think I've understood how it works, still practising programming in C, wasn't sure at first were argv came from, but that would be the command line arguments, right?

**Comment by CurlyMoo** [comment]
http://www.cprogramming.com/tutorial/c/lesson14.html

**Comment by pilino1234** [comment]
Still working on this, trying to implement reading the following options if there is a -F argument, but that seems to be going all right. 
I was just wondering whether the usage instructions on line 131 of the original file are correct, as the -l and -d options don't exist for -receive, and it looks more like -control usage to me. Should I fix that while I'm at it?

**Comment by CurlyMoo** [comment]
Don't see them...

**Comment by pilino1234** [comment]
In the printf() statement

**Comment by CurlyMoo** [comment]
Aha, indeed.

**Comment by pilino1234** [comment]
Should I change that?
Also, I can't figure out how to define the option -F accept string arguments. Any ideas?

**Comment by CurlyMoo** [comment]
Yes
`options_add(&options, 'F', "filter", OPTION_HAS_VALUE, 0, JSON_STRING, NULL, NULL);`

**Comment by pilino1234** [comment]
Ah, thanks :smiley:

**Comment by pilino1234** [comment]
I've nearly gotten it working now, but I can't figure out the logic when printing. I have this code:
[code]
json_find_string(jcontent, "protocol", &protocol);
for(j=0;j<m;j++) {
	printf("line 207: %d\n", j);
	if(strcmp(protocol, filters[j]) != 0) {
		char *content = json_stringify(jcontent, "\t");
		printf("%s\n", content);
		json_free(content);
	} else {
		continue;
	}
}
[/code]
However this filters out the first protocol which is listed in the arguments, but the second one still shows up. Do you have an idea on how I can make that work?

**Comment by CurlyMoo** [comment]
Can you show more surrounding code?

**Comment by pilino1234** [comment]
Yes:
```
	while(main_loop) {
		if(socket_read(sockfd, &recvBuff, 0) != 0) {
			goto close;
		}
		char *protocol = NULL;
		char **array = NULL;
		char **filters = NULL;
		unsigned int n = explode(recvBuff, "\n", &array), i = 0;
		int m = explode(filter, ",", &filters), j = 0;

		for(i=0;i<n;i++) {
			struct JsonNode *jcontent = json_decode(array[i]);
			struct JsonNode *jtype = json_find_member(jcontent, "type");
			if(jtype != NULL) {
				json_remove_from_parent(jtype);
				json_delete(jtype);
			}
			if(filteropt == 1) {
				json_find_string(jcontent, "protocol", &protocol);
				for(j=0;j<m;j++) {
					printf("line 207: %d\n", j);
					if(strcmp(protocol, filters[j]) != 0) {
						char *content = json_stringify(jcontent, "\t");
						printf("%s\n", content);
						json_free(content);
					} else {
						continue;
					}
				}
				array_free(&filters, m);
			} else {
				char *content = json_stringify(jcontent, "\t");
				printf("%s\n", content);
				json_free(content);
			}
			json_delete(jcontent);
		}
		array_free(&array, n);
	}
```
It's the entire main_loop that prints the data to output. Also see the update command above

**Comment by CurlyMoo** [comment]
1. Don't explode the filters in the main loop. You only need to initialize the filters once and can reuse them everytime.
2. You should free the filters array when m > 0.
3. You should do something like this:
```
int filtered = 0;
for(....) {
   if(strcmp(filter[..], protocol) == 0) {
       filtered = 1;
       break;
   }
}
if(filtered == 0) {
     printf(....);
}
[/code]

**Comment by pilino1234** [comment]
1. I've done that now
2. "You should free the filters array when m > 0." I would do that with ```array_free(&filters, m);```, right? But I'm not sure where I should put that...
3. Done that too, now I just need to do the array_free part

**Comment by CurlyMoo** [comment]
2. On all places that are required to make sure the array is freed. Currently, in cased where the `filteropt` is 0 the explode will still happend but is not freed.

**Comment by pilino1234** [comment]
So I can free it at the end of the mainloop when ```array``` is also freed (second to last line in the snippet above)?

**Comment by CurlyMoo** [comment]
Or move the explode so it actually makes sense to run it (when filteropt == 1).

**Comment by pilino1234** [comment]
Seems to work now:
```
	while(main_loop) {
		if(socket_read(sockfd, &recvBuff, 0) != 0) {
			goto close;
		}
		char *protocol = NULL;
		char **array = NULL;
		int filtered = 0;
		unsigned int n = explode(recvBuff, "\n", &array), i = 0;

		for(i=0;i<n;i++) {
			struct JsonNode *jcontent = json_decode(array[i]);
			struct JsonNode *jtype = json_find_member(jcontent, "type");
			if(jtype != NULL) {
				json_remove_from_parent(jtype);
				json_delete(jtype);
			}
			if(filteropt == 1) {
				char **filters = NULL;
				unsigned int m = explode(filter, ",", &filters), j = 0;
				json_find_string(jcontent, "protocol", &protocol);
				for(j=0;j<m;j++) {
					if(strcmp(filters[j], protocol) == 0) {
						filtered = 1;
						array_free(&filters, m);
						break;
					}
				}
				if(filtered == 0) {
					char *content = json_stringify(jcontent, "\t");
					printf("%s\n", content);
					json_free(content);
				} else {
					continue;
				}
			} else {
				char *content = json_stringify(jcontent, "\t");
				printf("%s\n", content);
				json_free(content);
			}
			json_delete(jcontent);
		}
		array_free(&array, n);
	}
```
Anything else I need to do?

**Comment by CurlyMoo** [comment]
Yes, run valgrind for memory leak checkes.

**Comment by pilino1234** [comment]
Ok, it's running now. Just let it run until it complains about something?

**Comment by CurlyMoo** [comment]
And ofc terminate with ctrl-c to see the final leakage report.

**Comment by pilino1234** [comment]
```
==8462== HEAP SUMMARY:
==8462==     in use at exit: 766 bytes in 67 blocks
==8462==   total heap usage: 14,145 allocs, 14,078 frees, 653,463 bytes allocated
==8462== 
==8462== 18 bytes in 1 blocks are definitely lost in loss record 1 of 4
==8462==    at 0x4C29110: malloc (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
==8462==    by 0x4C2B4AF: realloc (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
==8462==    by 0x414EC4: main (receive.c:135)
==8462== 
==8462== 748 (352 direct, 396 indirect) bytes in 22 blocks are definitely lost in loss record 4 of 4
==8462==    at 0x4C2B41E: realloc (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
==8462==    by 0x41D866: explode (common.c:138)
==8462==    by 0x415232: main (receive.c:205)
==8462== 
==8462== LEAK SUMMARY:
==8462==    definitely lost: 370 bytes in 23 blocks
==8462==    indirectly lost: 396 bytes in 44 blocks
==8462==      possibly lost: 0 bytes in 0 blocks
==8462==    still reachable: 0 bytes in 0 blocks
==8462==         suppressed: 0 bytes in 0 blocks
==8462== 
==8462== For counts of detected and suppressed errors, rerun with: -v
==8462== ERROR SUMMARY: 2 errors from 2 contexts (suppressed: 0 from 0)
```
2 places to fix...

**Comment by CurlyMoo** [comment]
Good luck!

**Comment by pilino1234** [comment]
```
==9002== 
==9002== HEAP SUMMARY:
==9002==     in use at exit: 0 bytes in 0 blocks
==9002==   total heap usage: 2,112 allocs, 2,112 frees, 99,815 bytes allocated
==9002== 
==9002== All heap blocks were freed -- no leaks are possible
==9002== 
==9002== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
==9002== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)```
I'll commit now :smiley:

**Comment by CurlyMoo** [comment]
Can you post the final loop?

**Comment by pilino1234** [comment]
```
while(main_loop) {
		if(socket_read(sockfd, &recvBuff, 0) != 0) {
			goto close;
		}
		char *protocol = NULL;
		char **array = NULL;
		unsigned int n = explode(recvBuff, "\n", &array), i = 0;

		for(i=0;i<n;i++) {
			struct JsonNode *jcontent = json_decode(array[i]);
			struct JsonNode *jtype = json_find_member(jcontent, "type");
			if(jtype != NULL) {
				json_remove_from_parent(jtype);
				json_delete(jtype);
			}
			if(filteropt == 1) {
				char **filters = NULL;
				unsigned int m = explode(filter, ",", &filters), j = 0;
				int filtered = 0;
				json_find_string(jcontent, "protocol", &protocol);
				for(j=0;j<m;j++) {
					if(strcmp(filters[j], protocol) == 0) {
						filtered = 1;
						break;
					}
				}
				array_free(&filters, m);
				if(filtered == 0) {
					char *content = json_stringify(jcontent, "\t");
					printf("%s\n", content);
					json_free(content);
				}
			} else {
				char *content = json_stringify(jcontent, "\t");
				printf("%s\n", content);
				json_free(content);
			}
			json_delete(jcontent);
		}
		array_free(&array, n);
	}
```

**Comment by CurlyMoo** [comment]
Indeed, that's what i also saw go wrong in the previous implementation :smile: But i knew you could figure it out yourself.

**Comment by pilino1234** [comment]
So I'll push now? No more improvements to be done? :smile:

**Comment by CurlyMoo** [comment]
When you want to do it perfectly, you can add a protocol validation check.

**Comment by pilino1234** [comment]
To make sure that only valid protocols are supplied? How can I get a list of all valid protocols in pilight?

**Comment by CurlyMoo** [comment]
Check pilight-send.

**Comment by pilino1234** [comment]
Yea, I was just going to edit and say that I might borrow some code from -control or -send

**Comment by pilino1234** [comment]
Hmm, I think I need to split the explode out of the main_loop and put it in an if switch to do the protocol validation there instead

**Comment by CurlyMoo** [comment]
Yes

**Comment by pilino1234** [comment]
I think I've gotten it to work, but I don't know how to get the value of m and j in the main_loop if I move this to somewhere else outside the main_loop:
```unsigned int m = explode(filter, ",", &filters), j = 0;```

**Comment by CurlyMoo** [comment]
Just do:
```
unsigned int m = 0, j = 0;
while(main_loop) {
   m = explode(filter, ",", &filters), j = 0;
}
```

**Comment by pilino1234** [comment]
But then I'll have the explode twice :/
I moved it to an earlier part for protocol validation...

**Comment by CurlyMoo** [comment]
ok
```
unsigned int nrfilter = explode(filter, ",", &filters), j = 0;
while(main_loop) {
   for(j=0;j<nrfilter;j++) {
   }
}
arrray_free(&filters, nrfilter);
```

**Comment by pilino1234** [comment]
Actually, never mind, I just put that definition ```unsigned int m = 0, j = 0;``` outside the protocol validation, and it doesn't complain about the variables any more, seems to work too

**Comment by pilino1234** [comment]
But it seems like I need to do ```char **filters = NULL;``` outside of the protocol validation too, in which case it will be defined too, even if -F isn't even used...

**Comment by CurlyMoo** [comment]
Yep. That doesn't matter. Defining a NULL variable uses hardly no resources.

**Comment by pilino1234** [comment]
Now it segfaults at ```pnode = pnode->next``` in this section (the protocol validation):
```
	if(filteropt == 1) {
		struct protocol_t *protocol = NULL;
		struct protocols_t *pnode = protocols;		
		unsigned int m = explode(filter, ",", &filters), j = 0;
		int match = 0;
		
		protocol_init();
		
		for(j=0;j<m;j++) {
			if(filters[j] != NULL && strlen(filters[j]) > 0) {
				while(pnode) {
					protocol = pnode->listener;
					if(protocol_device_exists(protocol, filters[j]) == 0 && match == 0 && protocol->createCode != NULL) {
						match = 1;
						break;
					}
				}
			}
			pnode = pnode->next;
		}
		if(match == 0) {
			printf("Invalid protocol specified");
			exit(EXIT_SUCCESS);
		}
	}
```

**Comment by CurlyMoo** [comment]
Because the `pnode = pnode->next` is not inside the correct loop.

**Comment by pilino1234** [comment]
True, just noticed. Too many brackets...

**Comment by CurlyMoo** [comment]
I've been working for over a week rewriting the arping functionality for the new major pilight rewrite. Thinkering for more then 5 minutes is not a sin :wink: In the mean while i learn how to program raw network communication.

**Comment by pilino1234** [comment]
Major pilight rewrite? Sounds nice :smiley:
What will change?

**Comment by CurlyMoo** [comment]
Please join IRC

**Comment by pilino1234** [comment]
Hmm, now it says that I specified an invalid protocol. Probably messed up the logic somewhere :stuck_out_tongue:

**Comment by pilino1234** [comment]
Ok, I think I've found the problem, but I don't know how to solve it. Basically ```while(pnode)``` doesn't seem to run, I guess because pnode isn't defined in a way that it likes....
```
	if(filteropt == 1) {
		struct protocol_t *protocol = NULL;
		struct protocols_t *pnode = protocols;		
		unsigned int m = explode(filter, ",", &filters), j = 0;
		int match = 0;
		
		protocol_init();
		
		for(j=0;j<m;j++) {
			printf("line 161\n");
			if(filters[j] != NULL && strlen(filters[j]) > 0) {
				printf("line 163: %s\n", filters[j]);
				while(pnode) {
					printf("line 165\n");
					protocol = pnode->listener;
					printf("line 167\n");
					if(protocol_device_exists(protocol, filters[j]) == 0 && match == 0 && protocol->createCode != NULL) {
						printf("line 169\n");
						match = 1;
						break;
					}
					printf("line 173\n");
					pnode = pnode->next;
				}
			}
		}
		if(match == 0) {
			printf("Invalid protocol specified\n");
			exit(EXIT_SUCCESS);
		}
	}
```
"line 165" never gets printed, 161 and 163 do. Here is the complete output:
```
line 161
line 163: datetime
line 161
line 163: alecto_wsd17
line 161
line 163: cpu_temp
Invalid protocol specified
```

**Comment by CurlyMoo** [comment]
The problem is that the pnode list can only run once because you never reset it to the beginning. That should be enough hints for now.

**Comment by pilino1234** [comment]
Ok, got it working now :smile:
I'll push so that you can have a look.

**Comment by pilino1234** [comment]
Should I make a new branch to only have one commit and make a PR from there?

**Comment by CurlyMoo** [comment]
Please read the contribution guide so see how to solve this without reopening the PR.

**Comment by pilino1234** [comment]
git reset --hard ?

**Comment by CurlyMoo** [comment]
This was the link:
http://eli.thegreenplace.net/2014/02/19/squashing-github-pull-requests-into-a-single-commit

**Comment by pilino1234** [comment]
Ok, I'll try to finish this tonight

**Comment by pilino1234** [comment]
I fixed the formatting for the first three comments, but I couldn't find a way to redefine m in the main_loop without redoing the filters. 
I pushed my attempt at doing it without redoing the filters to the receive_filter2 branch of my fork, but it segfaults at line 236, where it tries to compare the received protocol to one of the filters. I'm still trying to solve that.

**Comment by pilino1234** [comment]
Ok, fixed that too. Squashed and pushed :smiley:

File: receive.c [hunk 1]
```diff
@@ -34,6 +34,8 @@
 static int main_loop = 1;
 static int sockfd = 0;
 static char *recvBuff = NULL;
+char **filters = NULL;
```
Comment (CurlyMoo) [inline receive.c]: Why is filters defined twice?

File: receive.c [hunk 2]
```diff
@@ -127,14 +134,49 @@ int main(int argc, char **argv) {
...
+					protocol = pnode->listener;
+					if(protocol_device_exists(protocol, filters[j]) == 0 && match == 0) {
+						match = 1;
+						break;
+					}
+					pnode = pnode->next;
+				}
+				if(match == 0) {
+					logprintf(LOG_ERR, "Invalid protocol: %s", filters[j]);
+					exit(EXIT_SUCCESS);
```
Comment (CurlyMoo) [inline receive.c]: This prevents the garbage collection to run. Please use the `goto close`.

File: receive.c [hunk 3]
```diff
@@ -168,30 +210,50 @@ int main(int argc, char **argv) {
...
+        }
+        char *protocol = NULL;
+        char **array = NULL;
+        unsigned int n = explode(recvBuff, "\n", &array), i = 0;
+
+        for(i=0;i<n;i++) {
+            struct JsonNode *jcontent = json_decode(array[i]);
+            struct JsonNode *jtype = json_find_member(jcontent, "type");
+            if(jtype != NULL) {
+					json_remove_from_parent(jtype);
```
Comment (CurlyMoo) [inline receive.c]: Identation still doesn't look right.

File: receive.c (hunk 1 above)
Comment (pilino1234) [inline receive.c]: Dunno why GitHub says that this diff isn't outdated, removed the second definition.

**Comment by pilino1234** [comment]
New commit

**Pull request merged**